### ISSUE TITLE
S3 Aggregation Best Practice?

### ISSUE THREAD

**Issue by dblodgett-usgs** [Unidata/tds#215]
I'm testing with:
`unidata/thredds-docker:5.3`

I have a "**content root**":
`TDS_CONTENT_ROOT_PATH=/mnt/content`

And have mounted TDS configuration files there. Otherwise, I'm working in a stock TDS.

I have a `/mnt/thredds` that I use to store configuration files. I think of this as the "**data root**".

> Side note: 
> In our non-AWS production system, `/mnt/thredds` also houses all our netcdf data content, some of which are joinExisting aggregations of many thousands of files, some of which also union a collection of joinExisting aggregations such that we have many variables in one OPeNDAP endpoint. e.g. https://cida.usgs.gov/thredds/dodsC/loca_future.html

In my testing for migration to S3, I am trying to work out what the best configuration pattern will be to get performance out of these data.

I've determined that a catalog/ncml pattern like either of the following will function and result in a joinExisting cache that contains the indices of the joinExisting dimension:

```xml
  <dataset name="test prism agg regexp (works but SLOW?)" ID="prism-agg-3" urlPath="prism-agg/agg3" dataType="Grid" serviceName="all">
    <ncml:netcdf xmlns="http://www.unidata.ucar.edu/namespaces/netcdf/ncml-2.2" id="prism-agg-3">
      <aggregation dimName="time" type="joinExisting" timeUnitsChange="true">
        <scan location="cdms3://default@aws/nhgf-development?thredds/prism_v2/#delimiter=/" regExp=".*nc"/>
      </aggregation>
    </ncml:netcdf>
  </dataset>
```

or 

```xml
  <dataset name="test prism agg explicit" ID="prism-agg-2" urlPath="prism-agg/agg2" dataType="Grid" serviceName="all">
    <ncml:netcdf xmlns="http://www.unidata.ucar.edu/namespaces/netcdf/ncml-2.2" id="prism-agg-2">
      <aggregation dimName="time" type="joinExisting" timeUnitsChange="true">
        <netcdf location="cdms3://default@aws/nhgf-development?thredds/prism_v2/prism_2019.nc"/>
        <netcdf location="cdms3://default@aws/nhgf-development?thredds/prism_v2/prism_2020.nc"/>
      </aggregation>
    </ncml:netcdf>s
  </dataset>
```
Now to the crux of my issue. I'm getting very slow responses when querying these aggregations for coordinate variables which really need to be snappy. 

In contrast to the stand along catalog `dataset` elements above, I currently have my catalog and ncml layed out using three tiers that help separate concerns. 
1. xml THREDDS catalog metadata containing a `netcdf` element with a `location` attribute pointing to an ncml file. 
2. The ncml file contains dataset metadata and a `netcdf` element with a `location` attribute pointing to 
3. an `ncml` aggregation file with the contents of the xml snips above. 

My thought is to modify my ncml in the third tier so that, rather than pointing to data adjacent to it in a traditional file system, it points to a cdms3 location either using a scan or using explicit object paths. 

But then I still run up against this issue where my reads of coordinate variables are super slow. In this particular case it is because the time bounds variable is not cached and has to be retrieved to read in the file. In other cases, I would really like to avoid scanning all the data to populate a cache in the first place.

So I guess I have three questions. 

1. Is this three-tiered catalog metadata -> ncml metadata -> ncml aggregation layout going to cause us issues with caching or in other ways I'm not aware of? 
2. Is it possible to pre-cache coordinate variables in joinExisting or other ncml collections? Where would I go to find information about that?
3. long shot, but I've never used the TDM -- should I pursue deploying that for its utility in managing the caches?

Thanks -- and please tell me tho RTFM if one exists!

**Comment by haileyajohnson** [comment]
If only we had good enough documentation to just say RTFM :)

Overall, if you're finding S3 aggregations to be cumbersomely slow, that's _bad_ and we'd like to work with you on that. But first, we should try some aggregation caching options. Caching is on by default in the TDS, and caches coordinate variable for joinExisting collections the first time they are read (see [here](https://docs.unidata.ucar.edu/netcdf-java/current/userguide/ncml_aggregation.html#aggregation-caching).  You can configure your cache policy in your `threddsConfig.xml` (see [here](https://docs.unidata.ucar.edu/tds/current/userguide/tds_config_ref.html#aggregation-cache)). 

If your aggregations are dynamic, you'll also need to add a "recheckEvery" property to your aggregation.
I also think you need to change `regExp=".*nc"` to `suffix=".*nc"`.

Note that the first time accessing your datasets after starting your server might still be slow, but subsequent accesses _should_ be pretty fast. If they're not, we need to look into that.

To answer your other questions:
1. I'm not aware of any reason that layout would cause issues, but could I see what's going into the ncml metadata tier?
3. The TDM creates indexes for GRIB files, so it wouldn't help in your case, unfortunately

**Comment by dblodgett-usgs** [comment]
Thanks for following up @haileyajohnson -- I appreciate it.

My aggregation cache does appear to be working. I've set `scour` and `maxAge` to 
"-1" and I'm seeing the expected slow initial scan then fast response.

The dataset in question is hosted on our [old server here.](https://cida.usgs.gov/thredds/catalog.html?dataset=cida.usgs.gov/prism_v2) 

It is really just illustrative of this case where I'd really like a bit more than the joinExisting dimension to be cached. In this case, the `time_bounds` is slow from S3 because it has to be read out of every file. But more generally, I'd really like to be able to cache all 1D and maybe 2D variables such that queries for coordinate variables are always fast. Is this possible? Could I inject them into NCML such that they would never need to be retrieve from object store in the first place? I might just try that and see what happens! 

Thanks for the TDM nuance - I'll keep ignoring that :)

**Comment by dblodgett-usgs** [comment]
OK -- thanks @cofinoa -- I think this mostly answers my question.

> to create and external .ncml where all coordinates, with its values and metadata explicit created and just data aggregation on main variables it's perform, avoiding to read each file.

Do you have a sample of how you're doing this? And can you confirm that THREDDS will not attempt to readd data that has already been declared in the ncml directly?

**Comment by dblodgett-usgs** [comment]
Got it. Thank you!

**Comment by cofinoa** [comment]
you can persist the resulting aggregation dataset to a external NCML and ...
> You can point to this .ncml from your catalog entry.
>
> If you need "dynamical" aggregations, then a external cron can rebuild the external NCML and in combination with the recheckEvery in the catalog entry.

**Comment by cofinoa** [comment]
@dblodgett-usgs after the TDS start ... and the  first access to the Union aggregation based on `scan` .... the second and so on ... are timing out? 

From your URL example, my guess is that TDS has cached the aggregation after the first access, and the following access are much faster. 

can you share the NCML used in your example? the S3 endpoints are public accesible?

**Comment by dblodgett-usgs** [comment]
The first access times out and it is much faster while the aggregation is "fresh". But the result of the scan doesn't seem to be persisted any where so when the server is restarted or enough time and/or other usage has elapsed, the end point is subject to time out due to a rescan. I'm trying to figure out if there is a way to avoid this scan in the first place.

**Comment by cofinoa** [comment]
there is a config parameter related with aggregation cache:
https://docs.unidata.ucar.edu/tds/current/userguide/tds_config_ref.html#aggregation-cache

but It say that it's only for JoinExisting aggregations, but you can give a try.

**Comment by dblodgett-usgs** [comment]
I've verified that the aggregation cache does not work. 

Out of curiosity, I used toolsui to generate an ncml template of the dataset with `location="dods://..."` 

If I remove the `location` attribute and include the union aggregation within the `netcdf` element, I thought I was seeing a speed up of that initial scan, but upon further testing, it's not a panacea. The initial scan for some of these is still far too long.
